### PR TITLE
Jacobi on the KSP side

### PR DESCRIPTION
**Pull request by eggrobin**
#631 now occurs deterministically.
```
Apocalypse occurred at +3.67200000000000000e+05 s
```
Note that since KSP assumes that the orbital elements of the CelestialBodies never change (whereas principia reassigns them at every frame to put the bodies where they belong) we need to save the unmodified elements in a static object before we do anything.

### CODE REVIEW

File: ksp_plugin/plugin.cpp [hunk 1]
```diff
@@ -79,96 +79,108 @@ Plugin::Plugin(Instant const& initial_time,
...
-  DirectlyInsertCelestial(celestial_index,
-                          &parent_index,
-                          parent_degrees_of_freedom + relative,
-                          std::move(body));
-}
+// The three celestial-inserting functions log at the INFO level.  This is not
+// a concern since they are only called at initialization, which should happen
+// once per game.
 
 void Plugin::InsertSun(Index const celestial_index,
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: It's unclear if this function should still exist. It's now only usable for hierarchical initialization. Maybe the parent index should be optional in the hierarchical case and this function can go away. Leave a to-do?

File: ksp_plugin/plugin.cpp [hunk 2]
```diff
@@ -734,6 +746,41 @@ Plugin::Plugin(GUIDToOwnedVessel vessels,
...
+  absolute_initialization_ = std::experimental::nullopt;
+  ephemeris_ = std::make_unique<Ephemeris<Barycentric>>(
+      std::move(bodies),
+      initial_state,
+      current_time_,
+      history_integrator_,
+      kStep,
+      kFittingTolerance);
+  for (auto const& pair : celestials_) {
+    auto& celestial = *pair.second;
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: Const?

File: ksp_plugin/plugin.hpp [hunk 3]
```diff
@@ -337,6 +330,10 @@ class Plugin {
          Instant history_time,
          Index sun_index);
 
+  // We virtualize those tasks for testing purposes.
```
Comment (pleroy) [inline ksp_plugin/plugin.hpp]: Task should be functions

File: ksp_plugin/plugin.hpp [hunk 4]
```diff
@@ -90,23 +94,6 @@ class Plugin {
...
-  // |b.gravParameter|,
-  // |b.orbit.referenceBody.flightGlobalsIndex|,
-  // |{b.orbit.pos, b.orbit.vel}|.
-  virtual void InsertCelestial(
-    Index const celestial_index,
-    GravitationalParameter const& gravitational_parameter,
-    Index const parent_index,
-    RelativeDegreesOfFreedom<AliceSun> const& from_parent);
-
   // Inserts a celestial body with index |celestial_index| and gravitational
```
Comment (pleroy) [inline ksp_plugin/plugin.hpp]: Comment that this is only for hierarchical (I think)

File: ksp_plugin/plugin.hpp [hunk 5]
```diff
@@ -427,10 +424,25 @@ class Plugin {
...
 
-  // |bodies_| and |initial_state_| are null if and only if |!initializing_|.
-  // TODO(egg): optional.
-  std::unique_ptr<IndexToMassiveBody> bodies_;
-  std::unique_ptr<IndexToDegreesOfFreedom> initial_state_;
+
+  struct AbsoluteInitializationObjects {
+    IndexToMassiveBody bodies;
+    IndexToDegreesOfFreedom initial_state;
+  };
```
Comment (pleroy) [inline ksp_plugin/plugin.hpp]: Blank line after

File: ksp_plugin/plugin.cpp (hunk 2 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: `set_trajectory` is not const.

File: ksp_plugin/plugin.hpp (hunk 5 above)
Comment (pleroy) [inline ksp_plugin/plugin.hpp]: If like to remove the previous comment but I don't know how.

Comment (pleroy) [inline ksp_plugin/plugin.hpp]: I'd like to remove the previous comment but I don't know how.

Comment (pleroy) [inline ksp_plugin/plugin.hpp]: If you find these comments confusing you're not alone.

File: ksp_plugin_adapter/ksp_plugin_adapter.cs [hunk 6]
```diff
@@ -1287,17 +1308,29 @@ where is_in_physics_bubble(vessel)
       }
     } else {
       plugin_source_ = PluginSource.ORBITAL_ELEMENTS;
-      plugin_ = Interface.NewPlugin(Planetarium.GetUniversalTime(),
+      plugin_ = Interface.NewPlugin(0,
```
Comment (pleroy) [inline ksp_plugin_adapter/ksp_plugin_adapter.cs]: Comment why the change

File: ksp_plugin_test/interface_test.cpp [hunk 7]
```diff
@@ -250,28 +243,6 @@ TEST_F(InterfaceTest, DeletePlugin) {
   EXPECT_THAT(plugin, IsNull());
 }
 
-TEST_F(InterfaceTest, InsertCelestial) {
```
Comment (pleroy) [inline ksp_plugin_test/interface_test.cpp]: There should be a simple test for the Kepler insert

File: ksp_plugin_test/plugin_test.cpp [hunk 8]
```diff
@@ -345,56 +375,31 @@ TEST_F(PluginTest, Initialization) {
...
-                    AlmostEquals(looking_glass_.Inverse()(
-                            plugin_->CelestialFromParent(index).displacement()),
-                        1, 42380),
-                    AlmostEquals(looking_glass_.Inverse()(
-                            plugin_->CelestialFromParent(index).velocity()),
-                        74, 1475468))) << SolarSystemFactory::name(index);
+                    AlmostEquals(to_icrf(plugin_->CelestialFromParent(index)
+                                             .displacement()),
+                                 0,
+                                 42380),
```
Comment (pleroy) [inline ksp_plugin_test/plugin_test.cpp]: Keep the bounds on a single line

Comment (pleroy) [inline ksp_plugin_test/plugin_test.cpp]: Keep the bounds on a single line. Several places in this file.

File: physics/kepler_orbit.hpp [hunk 9]
```diff
@@ -1,7 +1,9 @@
 ﻿
 #pragma once
 
-#include "experimental/optional"
```
Comment (pleroy) [inline physics/kepler_orbit.hpp]: Hmmmm...

File: physics/kepler_orbit_body.hpp [hunk 10]
```diff
@@ -29,6 +32,36 @@ using quantities::Time;
...
+  };
+  append("e", elements.eccentricity);
+  if (elements.semimajor_axis) {
+    append("a", *elements.semimajor_axis);
+  }
+  if (elements.mean_motion) {
+    append("n", *elements.mean_motion);
+  }
+  append("i", elements.inclination);
+  append(u8"Ω", elements.longitude_of_ascending_node);
```
Comment (pleroy) [inline physics/kepler_orbit_body.hpp]: Do we have to say u8?

File: quantities/parser_body.hpp [hunk 11]
```diff
@@ -114,6 +114,11 @@ inline Angle ParseUnit(std::string const& s) {
 }
 
 template<>
+inline AngularFrequency ParseUnit(std::string const& s) {
```
Comment (pleroy) [inline quantities/parser_body.hpp]: Why inline?

**Comment by pleroy** [comment]
Not lgtm yet, I want to see the changes

File: quantities/parser_body.hpp (hunk 11 above)
Comment (eggrobin) [inline quantities/parser_body.hpp]: Because `template<>`, so there would be ODR issues.

File: physics/kepler_orbit_body.hpp (hunk 10 above)
Comment (eggrobin) [inline physics/kepler_orbit_body.hpp]: I think so, we had issues "°" before.:w

File: ksp_plugin_test/plugin_test.cpp (hunk 8 above)
Comment (eggrobin) [inline ksp_plugin_test/plugin_test.cpp]: Done.

File: ksp_plugin_test/interface_test.cpp (hunk 7 above)
Comment (eggrobin) [inline ksp_plugin_test/interface_test.cpp]: Done.

File: ksp_plugin_adapter/ksp_plugin_adapter.cs (hunk 6 above)
Comment (eggrobin) [inline ksp_plugin_adapter/ksp_plugin_adapter.cs]: Done.

File: ksp_plugin/plugin.hpp (hunk 5 above)
Comment (eggrobin) [inline ksp_plugin/plugin.hpp]: So do you want a blank line after or not??

Comment (eggrobin) [inline ksp_plugin/plugin.hpp]: I'd very much like the whole thing to be a variable of an unnamed struct, but then it wouldn't be optional (you can't declare a struct within a template parameter).

File: ksp_plugin/plugin.hpp (hunk 4 above)
Comment (eggrobin) [inline ksp_plugin/plugin.hpp]: Done.

File: ksp_plugin/plugin.hpp (hunk 3 above)
Comment (eggrobin) [inline ksp_plugin/plugin.hpp]: Done.

File: ksp_plugin/plugin.cpp (hunk 1 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: Well, some struct containing elements and parent index should be optional, but yes.
